src/Microsoft.AspNetCore.Razor.Language/Components/ComponentDocumentClassifierPass.cs

```diff
@@ -0,0 +1,133 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Text;
+using Microsoft.AspNetCore.Razor.Language.Intermediate;
+
+namespace Microsoft.AspNetCore.Razor.Language.Components
+{
+    internal class ComponentDocumentClassifierPass : DocumentClassifierPassBase
+    {
+        public static readonly string ComponentDocumentKind = "component.1.0";
+        private static readonly object BuildRenderTreeBaseCallAnnotation = new object();
+        private static readonly char[] PathSeparators = new char[] { '/', '\\' };
+        private static readonly char[] NamespaceSeparators = new char[] { '.' };
+
+        protected override string DocumentKind => ComponentDocumentKind;
+
+        protected override bool IsMatch(RazorCodeDocument codeDocument, DocumentIntermediateNode documentNode) => true;
+
+        protected override void OnDocumentStructureCreated(RazorCodeDocument codeDocument, NamespaceDeclarationIntermediateNode @namespace, ClassDeclarationIntermediateNode @class, MethodDeclarationIntermediateNode method)
+        {
+            base.OnDocumentStructureCreated(codeDocument, @namespace, @class, method);
+
+            if (!TryComputeNamespaceAndClass(
```

Comment:
Copied most of this method as is. I filed a follow up to figure out the base namespace from the project: https://github.com/aspnet/Razor/issues/2735